DisablePluginException now can be used to disable plugins

JavierLeon9966 opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Currently there is no way to explicitly disable a plugin, and the only way just causes boilerplate.
### Relevant issues
<!-- List relevant issues here -->
<!--

* Fixes #1
* Fixes #2

-->
* Fixes #2671
## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
* `DisablePluginException` can be used to disable a plugin when thrown inside `Plugin->onEnable()`
## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
```php
<?php

declare(strict_types = 1);

namespace JavierLeon9966\PluginExceptionTest;

use pocketmine\plugin\DisablePluginException;
use pocketmine\plugin\PluginBase;

class Main extends PluginBase{
    public function onEnable(): void{
        throw new DisablePluginException("Testing!");
    }
}
```

JavierLeon9966 committed:
PluginException now can be used to disable plugins

dktapps commented:
PluginException is too generic. It's thrown for error cases which should crash the server, such as attempting to register an event handler with an invalid priority.

JavierLeon9966 commented:
What about DisablePluginException? If so then what exception should it extend?

dktapps commented:
It doesn't really matter what it extends. RuntimeException would be fine.

NB: I already attempted to implement this once and there were issues that I don't remember. It's not as simple as it looks.

JavierLeon9966 committed:
Fixed incorrect argument being passed

JavierLeon9966 committed:
Changed PluginException to DisablePluginException for disabling plugins

SOF3 commented:
What is the point of *disabling* the plugin instead of crashing the server?

dktapps commented:
Crashing indicates a bug. A plugin may want to disable itself for non-bug reasons, such as failure to connect to a database, invalid configuration, or other reasons. None of these things are bugs in the code, so they shouldn't generate crashdumps.

SOF3 commented:
In those cases they should still stop the server. While not generating a crashdump, it does not make sense to continue running the server with a possibly critical plugin disabled automatically.

dktapps commented:
> In those cases they should still stop the server. While not generating a crashdump, it does not make sense to continue running the server with a possibly critical plugin disabled automatically.

This has already been discussed multiple times and it's outside the scope of this issue.

dktapps committed:
Merge branch 'next-minor' into issues/2671

dktapps commented on src/plugin/PluginManager.php:
```diff
@@ -434,7 +434,17 @@ public function enablePlugin(Plugin $plugin) : void{
+				$errorMessage = $e->getMessage();
+				if($errorMessage !== ""){
+					$plugin->getLogger()->error($errorMessage);
+				}
```
I think it would be better to omit this, and let the plugin itself do any error logging, so that it can localize any messages as needed. Passing a message to DisablePluginException doesn't make much sense since you can't translate.

SOF3 replied:
What prevents the plugin from providing a translated message to the exception?

dktapps replied:
They aren't able to inject language strings into any language context that the server knows about. It's better to just let them log the way they normally would.

SOF3 replied:
I said *translated* not *translatable*

dktapps replied:
My bad, did not see that.

JavierLeon9966 committed:
Do not log DisablePluginException's error message

This will let plugins do their own logging that could allow translatable messages.

JavierLeon9966 committed:
Merge branch 'next-minor' into issues/2671

JavierLeon9966 committed:
Return false if the plugin disabled itself

JavierLeon9966 committed:
Broadcast that the plugin has disabled itself

dktapps committed:
final

dktapps committed:
Merge branch 'next-minor' into issues/2671
